### PR TITLE
Add war file support

### PR DESCRIPTION
**Pull request by jotadrilo**
This adds support for Jenkins WAR files.

Jenkins bundles some plugins in the WAR file (detached plugins) which are installed at boot time after the rest of plugins get installed. In order to have a fully-qualified list of plugins offline, we need to parse the WAR file to add these core dependencies and find any incompatibility with our project dependencies.

Use case:

```console
$ jpresolver -input plugins.jsonnet -war jenkins.war -optional
2019/10/02 17:26:01 adding Jenkins bundled plugin to the locked list: antisamy-markup-formatter:1.1
2019/10/02 17:26:01 adding Jenkins bundled plugin to the locked list: command-launcher:1.0
2019/10/02 17:26:01 adding Jenkins bundled plugin to the locked list: cvs:2.11
2019/10/02 17:26:01 adding Jenkins bundled plugin to the locked list: external-monitor-job:1.4
2019/10/02 17:26:01 overwriting locked plugin javadoc:1.0 with Jenkins bundled plugin javadoc:1.1
2019/10/02 17:26:01 adding Jenkins bundled plugin to the locked list: jaxb:2.3.0
2019/10/02 17:26:01 adding Jenkins bundled plugin to the locked list: jdk-tool:1.0
2019/10/02 17:26:01 adding Jenkins bundled plugin to the locked list: ldap:1.11
2019/10/02 17:26:01 adding Jenkins bundled plugin to the locked list: pam-auth:1.1
2019/10/02 17:26:01 adding Jenkins bundled plugin to the locked list: ssh-slaves:1.15
2019/10/02 17:26:01 adding Jenkins bundled plugin to the locked list: translation:1.10
2019/10/02 17:26:01 adding Jenkins bundled plugin to the locked list: windows-slaves:1.0
2019/10/02 17:26:01  There were found some incompatibilities within jenkins.war:
2019/10/02 17:26:01   ├── jenkins.war > ant:1.2 (a newer version was bundled in the Jenkins war, please update your project file):
2019/10/02 17:26:01   │   └── ant: 1.0 -> 1.2

```

### CODE REVIEW

File: cmd/jpresolver/main.go [hunk 1]
```diff
@@ -110,6 +144,9 @@ func validateFlags() error {
 	if err := meta.EnsureStorePathExists(*workingDir); err != nil {
 		return errors.Trace(err)
 	}
+	if err := war.EnsureStorePathExists(*workingDir); err != nil {
```
Comment (migmartri) [inline cmd/jpresolver/main.go]: I am wondering if ensureStorePathExists can be in `common` and just pass the different path generators, i,e `GetWarPath`

File: cmd/jpresolver/main.go [hunk 2]
```diff
@@ -70,9 +73,18 @@ func readInput() (*api.Project, error) {
 
 func writeOutput(pr *api.PluginsRegistry) error {
 	outputFile := fmt.Sprintf("%s-lock.%s", strings.TrimSuffix(*inputFile, filepath.Ext(*inputFile)), "json")
+	sort.Sort(api.ByName(pr.Plugins))
 	return utils.MarshalJSON(outputFile, pr)
 }
 
+func readWar() (*api.PluginsRegistry, error) {
```
Comment (migmartri) [inline cmd/jpresolver/main.go]: maybe having this function is not that relevant. It relies on the global scope and in terms of LOC it's pretty much the same than it's call statement.

File: pkg/plugins/war/incompatibilities.go [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+	"log"
+
+	api "github.com/bitnami-labs/jenkins-plugins-resolver/api"
+	"github.com/bitnami-labs/jenkins-plugins-resolver/pkg/plugins/common"
+	"github.com/bitnami-labs/jenkins-plugins-resolver/pkg/utils"
+	"github.com/juju/errors"
+	"github.com/mmikulicic/multierror"
+)
+
+// AggregateBundledPlugins walks through a jenkins bundled plugins locked registry and compare their versions with the inputs locked registry.
```
Comment (migmartri) [inline pkg/plugins/war/incompatibilities.go]: Oh wow, thanks for the documentation.

File: pkg/plugins/war/extractor.go [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+
+	return string(data), nil
+}
+
+// ExtractDetachedPluginsManifests takes a war and returns a reader to a detached plugin manifest content
+func ExtractDetachedPluginsManifests(war string, detachedPlugins []string) (map[string]string, error) {
+	detachedPluginsBytes, err := zipper.ExtractFiles(war, detachedPlugins)
```
[the remark this answers is not in the document]
Comment (jotadrilo) [reply pkg/plugins/war/extractor.go]: Yes, I am afraid it is a tradeoff between opening/uncompressing the war file several times or keeping them in RAM.

**Pull request merged**